client/beefy/src/worker.rs

```diff
@@ -216,77 +241,225 @@ where
+				.expect("forwards closure result; the closure always returns Ok; qed.")
+		});
+		// Set new best BEEFY block number.
+		self.best_beefy_block = Some(block_num);
+		metric_set!(self, beefy_best_block, block_num);
+	}
+
+	/// Return first block number pertaining to same session as `header`.
+	fn session_start_for_header(&self, header: &B::Header) -> NumberFor<B> {
+		// Current header is also the session start.
+		if let Some(_) = find_authorities_change::<B>(header) {
+			return *header.number()
+		}
+
+		// Walk up the chain looking for the session change digest.
+		let mut parent_hash = *header.parent_hash();
+		while parent_hash != Default::default() {
+			let header = self
+				.client
+				.expect_header(BlockId::Hash(parent_hash))
+				// TODO: is this proof correct?
+				.expect("header always available when following parent hash; qed.");
+			if let Some(_) = find_authorities_change::<B>(&header) {
+				// Found the first header of the session.
+				break
+			}
+			parent_hash = *header.parent_hash();
+		}
+		*header.number()
+	}
+
+	/// Handle potential session changes by starting new voting round for mandatory blocks.
+	fn handle_session_change(&mut self) {
+		let header = &self.best_grandpa_block_header;
+		if let Some(active) = self.validator_set(header) {
+			// If validator set has changed compared to the one in our active round,
 			if self.rounds.is_none() ||
-				active.id() != self.rounds.as_ref().unwrap().validator_set_id() ||
-				(active.id() == GENESIS_AUTHORITY_SET_ID && self.best_beefy_block.is_none())
+				active.id() != self.rounds.as_ref().unwrap().validator_set_id()
 			{
-				debug!(target: "beefy", "🥩 New active validator set id: {:?}", active);
-				metric_set!(self, beefy_validator_set_id, active.id());
+				// Find new session_start and start new round for mandatory block.
+				let session_start = self.session_start_for_header(header);
 
+				debug!(target: "beefy", "🥩 New active validator set: {:?}", active);
+				metric_set!(self, beefy_validator_set_id, active.id());
 				// BEEFY should produce a signed commitment for each session
-				if active.id() != self.last_signed_id + 1 && active.id() != GENESIS_AUTHORITY_SET_ID
+				if active.id() != GENESIS_AUTHORITY_SET_ID && active.id() != self.last_signed_id + 1
 				{
 					metric_inc!(self, beefy_skipped_sessions);
 				}
 
 				if log_enabled!(target: "beefy", log::Level::Debug) {
 					// verify the new validator set - only do it if we're also logging the warning
-					let _ = self.verify_validator_set(notification.header.number(), &active);
+					let _ = self.verify_validator_set(header.number(), &active);
 				}
 
 				let id = active.id();
-				self.rounds = Some(round::Rounds::new(active));
+				self.rounds = Some(round::Rounds::new(session_start, active));
+				debug!(target: "beefy", "🥩 New Rounds for validator set id: {:?} with session_start {:?}", id, session_start);
+			}
+		} else {
+			// TODO: BEEFY pallet not available, stop beefy worker
```

Comment:
Not necessarily stop, we should rather wait until the pallet becomes available (via runtime upgrade).